Doc/library/gc.rst

```diff
@@ -33,6 +33,34 @@ The :mod:`gc` module provides the following functions:
    Disable automatic garbage collection.
 
 
+.. class:: ensure_disabled()
+
+   Return a context manager object that disables the garbage collector and reenables the previous
+   state upon completion of the block. This is basically equivalent to::
+
+     from gc import enable, disable, isenabled
+
+     @contextmanager
+     def ensure_disabled():
+         was_enabled_previously = isenabled()
+         gc.disable()
+         yield
+         if was_enabled_previously:
+             gc.enable()
+
+   And lets you write code like this::
+
+     with ensure_disabled():
+         run_some_timing()
+
+     with ensure_disabled():
+         # do_something_that_has_real_time_guarantees
+         # such as a pair trade, robotic braking, etc
+
+   without needing to explicitly enable and disable the garbage collector yourself.
+   This context manager is implemented in C to assure atomicity, thread safety and speed.
```

Comment:
It would be good to clarify what atomicity and thread safety is assured (if any). I understand there is just one GC setting that applies to all threads.

What happens if this context manager is used in two threads at once? In particular, if thread B exiting the _with_ statement can enable GC when thread A is still inside a _with_ statement, that does not seem “safe”. What is the effect of calling _gc.disable_ while another thread runs the context manager? Is GC re-enabled at `__exit__`, or is it disabled indefinitely?

Reply:
I will complement the documentation explaining this. Just to clarify the current behaviour: The context manager deactivates the garbage collector and restores the previous state of the garbage collector in the __exit__ method. So if one thread uses the context manager while other has called it already, the garbage collector will be kept deactivated at the exit of the second invocation and it will be reactivated only at the end of the first (assuming that it was activated initially). So the guarantee is: the garbage collector will be deactivated at the entering of the context manager and the initial state will be restored at the end.

Reply:
As far as thread-safety goes: the CM is just as thread-unsafe as doing this dance manually. That means if thread A enters first, then thread B enters, then thread A exits, the gc *will* be turned back on even though thread B is still inside the context manager.

While we could choose to make this more thread-safe by replacing the CM-local "Was the GC enabled?" flag with a process-global counted semaphore:

* that still isn't entirely thread-safe due to the possibility of direct calls to the underlying APIs
* it adds pointless overhead to the single-threaded case

What could be useful though is a "gc.isenabled()" check in __exit__ that emits RuntimeWarning if the GC is already enabled at that point. That way, while the CM wouldn't magically fix your thread safety problems, it *would* help you find out you had them in the first place.

Reply:
+1 to the `gc.isenabled()`. Should I proceed to implement it?

Reply:
Notice that this still have edge cases for example:

1. Thread A enters (gc disabled)
2. Thread B enters (gc disabled)
3. Thread A exits (gc enabled) <-- bad
4. Thread C enters (gc disabled)
5. Thread C exists (gc disabled)
6. Thread B exists (gc disabled )

In this example, the hypothetical RuntimeWarning will never get triggered and thread B will never know that is has been running with the gc disabled for a while.

Reply:
Right, the documentation should still say the context manager is *not* thread-safe, and can't be readily made so due to existing code that calls the low-level APIs directly.

https://www.python.org/dev/peps/pep-0556/ is likely to be a better way forward for handling GC pauses in a thread-safe way than ad hoc disabling of the GC around performance critical bits of code.

Reply:
@ncoghlan In 00222b4 I have implemented a way to emit a warning if the garbage collector is activated when some thread is inside the `gc.ensure_disabled` context. Please, review it and if its ok I will update the documentation with the current behaviour and caveats. Of course if you prefer a different implementation I am more than happy to change it :smile:

Reply:
Notice that this implementation is really one step away from the counted semaphore and it covers the public API so maybe it makes sense to completely disallow enabling the gc if another thread is in inside the context. A microbenchmark reveals that is only 5.47% +- 3.2 (mean -+ stdv) slower than the previous one.